### PR TITLE
add destroy to release immediately for  native object

### PR DESCRIPTION
**Pull request by bofeng-song**
related pr:
https://github.com/cocos-creator-packages/jsb-adapter/pull/376
https://github.com/cocos-creator-packages/jsb-adapter/issues/370
https://github.com/cocos/engine-native/pull/4307

### CODE REVIEW

File: cocos/editor-support/spine-creator-support/SkeletonAnimation.h [hunk 1]
```diff
@@ -92,6 +92,7 @@ class SkeletonAnimation: public SkeletonRenderer {
 
     virtual void onAnimationStateEvent (TrackEntry* entry, EventType type, Event* event);
     virtual void onTrackEntryEvent (TrackEntry* entry, EventType type, Event* event);
+    void destroy() override;
```
Comment (minggo) [inline cocos/editor-support/spine-creator-support/SkeletonAnimation.h]: It is better to use virtual as other functions.

File: cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp [hunk 2]
```diff
@@ -146,22 +146,28 @@ SkeletonRenderer::SkeletonRenderer (const std::string& skeletonDataFile, const s
...
+    if (_ownsSkeletonData) {
+        delete _skeleton->getData();
+        _ownsSkeletonData = false;
     }
-    
-    CC_SAFE_RELEASE(_attachUtil);
-    CC_SAFE_RELEASE(_nodeProxy);
-    CC_SAFE_RELEASE(_effect);
+    CC_SAFE_DELETE(_skeleton);
+    CC_SAFE_DELETE(_atlas);
```
Comment (dumganhar) [inline cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp]:
```suggestion
    if (_ownAtlas) {
        CC_SAFE_DELETE(_atlas);
    }
```

File: cocos/editor-support/spine-creator-support/SkeletonAnimation.cpp [hunk 3]
```diff
@@ -139,9 +143,12 @@ SkeletonAnimation::~SkeletonAnimation () {
 
     if (_state) {
         clearTracks();
-        if (_ownsAnimationStateData) delete _state->getData();
-        delete _state;
+        if (_ownsAnimationStateData) {
+            delete _state->getData();
```
Comment (dumganhar) [inline cocos/editor-support/spine-creator-support/SkeletonAnimation.cpp:147]:
```suggestion
            if (_state != nullptr) {
                delete _state->getData();
            }
```

File: cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp [hunk 4]
```diff
@@ -150,16 +150,23 @@ SkeletonRenderer::~SkeletonRenderer () {
...
     if (_ownsSkeleton) {
         CC_SAFE_DELETE(_skeleton);
+    } else {
+        _skeleton = nullptr;
     }
     if (_ownsAtlas && _atlas) {
         CC_SAFE_DELETE(_atlas);
+    } else {
+        _atlas = nullptr;
     }
```
Comment (minggo) [inline cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp:166]: May be it is better to modify all codes like this
```c++
if (_ownsSkeleton) {
    delete(_skeleton);
}
if (_ownsAtlas) {
    delete _atlas;
}
...

_skeleton = nullptr;
_atlas = nullptr;
...
```

- no matter need to delete the instance, the pointer should be set to null.
- delete a null pointer is not a problem.

File: cocos/editor-support/spine-creator-support/SkeletonAnimation.cpp (hunk 3 above)
Comment (bofeng-song) [reply cocos/editor-support/spine-creator-support/SkeletonAnimation.cpp:147]: _state has been check at line 144

Comment (dumganhar) [reply cocos/editor-support/spine-creator-support/SkeletonAnimation.cpp:147]: ok

**Review by dumganhar — APPROVED** [review]

**Review by minggo — APPROVED** [review]

**Pull request merged**